Mathlib/Topology/MetricSpace/Kuratowski.lean

```diff
@@ -112,21 +110,77 @@ end KuratowskiEmbedding
   Classical.choose (KuratowskiEmbedding.exists_isometric_embedding α)
 #align Kuratowski_embedding kuratowskiEmbedding
 
-/-- The Kuratowski embedding is an isometry. -/
+/--
+The Kuratowski embedding is an isometry.
+Theorem 2.1 of [Assaf Naor, *Metric Embeddings and Lipschitz Extensions*][Naor-2015]. -/
 protected theorem kuratowskiEmbedding.isometry (α : Type u) [MetricSpace α] [SeparableSpace α] :
     Isometry (kuratowskiEmbedding α) :=
   Classical.choose_spec (exists_isometric_embedding α)
 #align Kuratowski_embedding.isometry kuratowskiEmbedding.isometry
 
 /-- Version of the Kuratowski embedding for nonempty compacts -/
 nonrec def NonemptyCompacts.kuratowskiEmbedding (α : Type u) [MetricSpace α] [CompactSpace α]
-    [Nonempty α] : NonemptyCompacts ℓ_infty_ℝ where
+    [Nonempty α] : NonemptyCompacts ℓ^∞(ℕ) where
   carrier := range (kuratowskiEmbedding α)
   isCompact' := isCompact_range (kuratowskiEmbedding.isometry α).continuous
   nonempty' := range_nonempty _
 #align nonempty_compacts.Kuratowski_embedding NonemptyCompacts.kuratowskiEmbedding
+
+/--
+A function `f : α → ℓ^∞(ι, ℝ)` which is `K`-Lipschitz on a subset `s` admits a `K`-Lipschitz
+extension to the whole space.
+
+Theorem 2.2 of [Assaf Naor, *Metric Embeddings and Lipschitz Extensions*][Naor-2015]
+
+The same result for the case of a finite type `ι` is implemented in
+`LipschitzOnWith.extend_pi`.
+-/
+theorem LipschitzOnWith.extend_lp_infty [PseudoMetricSpace α] {s : Set α} {f : α → ℓ^∞(ι)}
+    {K : ℝ≥0} (hfl : LipschitzOnWith K f s): ∃ g : α → ℓ^∞(ι), LipschitzWith K g ∧ EqOn f g s := by
+  -- Construct the coordinate-wise extensions
+  have : ∀ i : ι, ∃ g : α → ℝ, LipschitzWith K g ∧ EqOn (fun x => f x i) g s
+  · intro i
+    apply LipschitzOnWith.extend_real -- use the nonlinear Hahn-Banach theorem here!
+    rw [lipschitzOnWith_iff_dist_le_mul] at hfl ⊢
+    intro x hx y hy
+    calc
+      dist (f x i) (f y i) ≤ dist (f x) (f y) := lp.norm_apply_le_norm top_ne_zero (f x - f y) i
+      _ ≤ K * dist x y := hfl x hx y hy
+  choose g hg using this
+  rcases s.eq_empty_or_nonempty with rfl | ⟨a₀, ha₀_in_s⟩
+  . exact ⟨fun _ ↦ 0, LipschitzWith.const' 0, by simp⟩
+  · let f_ext : α → ι → ℝ := fun x i ↦ g i x
+    -- Show that the extensions are uniformly bounded
+    have hf_extb : ∀ a : α, Memℓp (f_ext a) ∞
+    · intro a
+      let M : ℝ := ‖f a₀‖
+      use K * dist a a₀ + M
+      rintro - ⟨i, rfl⟩
+      rcases hg i with ⟨hgl, hgr⟩
+      calc
+        |g i a| = |g i a - f a₀ i + f a₀ i| := by simp
+        _ ≤ |g i a - f a₀ i| + |f a₀ i| := abs_add _ _
+        _ ≤ |g i a - g i a₀| + |g i a₀ - f a₀ i| + |f a₀ i| := by gcongr; apply abs_sub_le
+        _ = |g i a - g i a₀| + |f a₀ i| := by simp [hgr ha₀_in_s]
+        _ ≤ ↑K * dist a a₀ + |f a₀ i| := by
+            gcongr
+            exact lipschitzWith_iff_dist_le_mul.1 hgl a a₀
+        _ ≤ ↑K * dist a a₀ + M := by
+            gcongr
+            exact lp.norm_apply_le_norm top_ne_zero (f a₀) i
```

Comment:
```suggestion
      calc
        |g i a| = |g i a - g i a₀ + f a₀ i| := by rw [←hgr ha₀_in_s, sub_add_cancel]
        _ ≤ |g i a - g i a₀| + |f a₀ i| := abs_add _ _
        _ ≤ ↑K * dist a a₀ + M := by
            apply add_le_add
            . exact lipschitzWith_iff_dist_le_mul.1 hgl a a₀
            . exact lp.norm_apply_le_norm top_ne_zero (f a₀) i
```

Reply:
@tb65536 Personally I prefer to avoid calling lemmas like `sub_add_cancel` and `add_le_add` by name ...
```suggestion
      calc
        |g i a| = |g i a - g i a₀ + f a₀ i| := by simp [hgr ha₀_in_s]
        _ ≤ |g i a - g i a₀| + |f a₀ i| := abs_add _ _
        _ ≤ ↑K * dist a a₀ + M := by
            gcongr
            . exact lipschitzWith_iff_dist_le_mul.1 hgl a a₀
            . exact lp.norm_apply_le_norm top_ne_zero (f a₀) i
```

Reply:
I was scrolling down the page, saw Thomas' suggestion and thought: quick, I need to write a PM to Thomas before Heather sees this. But it was too late.

Reply:
Old habits die hard... I'm too used to golfing to term-mode.